src/switch_ivr_async.c

```diff
@@ -2144,6 +2144,101 @@ struct eavesdrop_pvt {
 	uint8_t *frame_data;
 };
 
+static void handle_replace_frame(switch_media_bug_t *bug, struct eavesdrop_pvt *ep, switch_core_session_t *session, switch_bool_t is_read)
+{
+	switch_buffer_t *buffer = is_read ? ep->r_buffer : ep->w_buffer;
+	uint32_t flag = is_read ? ED_MUX_READ : ED_MUX_WRITE;
+	const char *log_prefix = is_read ? "read" : "write";
+	const char *log_skip_prefix = is_read ? "READ_REPLACE" : "WRITE_REPLACE";
+	switch_frame_t *rframe;
+	uint32_t required_bytes;
+	void *data_to_merge = ep->data;
+	uint32_t merge_samples = 0;
+	uint32_t buffer_inuse;
+	uint32_t input_samples;
+	int channels;
+
+	if (!switch_test_flag(ep, flag)) {
+		return;
+	}
+
+	rframe = is_read ? switch_core_media_bug_get_read_replace_frame(bug) : switch_core_media_bug_get_write_replace_frame(bug);
+	if (!rframe) {
+		return;
+	}
+
+	channels = rframe->channels ? rframe->channels : 1;
+
+	/* Calculate how much data we need from buffer */
+	if (ep->reverse_resampler) {
+		/* Calculate input samples needed to produce rframe->samples output samples */
+		input_samples = (uint32_t)((double)rframe->samples * ep->reverse_resampler->rfactor);
+		required_bytes = input_samples * 2 * ep->read_impl.number_of_channels;
+	} else {
+		required_bytes = rframe->datalen;
+	}
+
+	buffer_inuse = switch_buffer_inuse(buffer);
+
+	if (buffer_inuse >= required_bytes) {
+		uint32_t bytes;
+
+		switch_buffer_lock(buffer);
+		bytes = (uint32_t) switch_buffer_read(buffer, ep->data, required_bytes);
+
+		/* Apply reverse resampling if needed */
+		if (ep->reverse_resampler && bytes > 0) {
+			int16_t *original_data = (int16_t *)ep->data;
+			int original_samples = bytes / 2 / ep->read_impl.number_of_channels;
+
+			switch_mutex_lock(ep->reverse_resample_mutex);
+			switch_resample_process(ep->reverse_resampler, original_data, original_samples);
+
+			/* Copy resampled data to reverse resample buffer */
+			{
+				uint32_t out_bytes = ep->reverse_resampler->to_len * 2 * ep->read_impl.number_of_channels;
+				if (out_bytes > SWITCH_MAX_L16) out_bytes = SWITCH_MAX_L16;
+				memcpy(ep->reverse_resample_data, ep->reverse_resampler->to, out_bytes);
+			}
+			data_to_merge = ep->reverse_resample_data;
```

Comment:
`data_to_merge` is a pointer to the `ep->reverse_resample_data` then this `data_to_merge` is being access outside of mutex. Then it is possible to be overwritten by the other read/write thread before the merge happen